### PR TITLE
Adds search functionality 

### LINKED ISSUE #111: Search functionality required
No search functionality exists in project yet

### PR DESCRIPTION
**Pull request by idan1356**
Adds basic free-text search functionality for 3 entity types- `User`, `Study Group`, and `Teacher Course`,
----
- Create function `search_users_by_keyword` to use for filtering users based on a search keyword and add test
- Add `course/one_course.html` template to represent a single course in search view, and respective `one_course.css`
- Create `search.html` template to display search result of users, study groups and teacher course
- Create view for search and its tests. 
- Add search to navbar 

----

Dependencies: #108 (merged)

Resolves #111 

----

Screenshot:
search_users
Screenshot 2023-06-06 101730
search_study_groups

### CODE REVIEW

**Comment by idan1356** [comment]
> @idan1356 If u want I can help with the implementation of the teacher course search

Appreciate the offer to help
If you want you can create a `one_course.html` template for displaying a single course similar to `one_user.html` in users app
if not no worries I can do it too, that's all there is to do regarding teacher course for this PR I think

File: static/styles/one_course.css [hunk 1]
```diff
@@ -0,0 +1,38 @@
...
+  }
+
+#one_course a{
+    text-decoration: none;
+    color: inherit;
+}
+
+#one_course h5 {
+    color: #2E8B57;
+  }
```
Comment (shirazyomtov) [inline static/styles/one_course.css:38]: Missing one new line at the end of the file

**Review by shirazyomtov — CHANGES_REQUESTED** [review]
Great work, I left a comment.
Also you have 5 commits, you should squash them.

File: static/styles/one_course.css (hunk 1 above)
Comment (idan1356) [reply static/styles/one_course.css]: fixed

**Comment by idan1356** [comment]
> Great work, I left a comment. Also you have 5 commits, you should squash them.

The PR was rebased from its dependency, resulting in four commits instead of five
Also, I tried to keep each commit small and focused on a single purpose, that's why the division into multiple commits.

**Comment by idan1356** [comment]
Resolved conflicts

**Review by Davidco94 — DISMISSED** [review]

File: main/templates/search.html [hunk 2]
```diff
@@ -0,0 +1,51 @@
...
+                <p>No users available.</p>
+                {% endfor %}
+            </div>
+          </div>
+          <div class="tab-pane fade" id="tab3">
+              <div class="container">
+                <div class="row">
+                    {% for course in teacher_courses %}
+                      {% include "course/one_course.html" %}
+                    {% endfor %}
```
[the remark this answers is not in the document]
Comment (idan1356) [reply main/templates/search.html]: Thanks! been trying to change it forever
fixed

**Review by RanyWh — DISMISSED** [review]
LGTM

**Review by natifridman — COMMENTED** [review]
This PR isn't draft anymore. Please edit the PR description accordingly

File: tests/main/test_search_page.py [hunk 3]
```diff
@@ -0,0 +1,60 @@
...
+    def test_search_context_equals_expected_queryset_study_groups(self, make_search_http_response_from_query_param,
+                                                                  search_query_param, study_group_context_name):
+        search_response = make_search_http_response_from_query_param(search_query_param)
+
+        expected_study_group_queryset = StudyGroup.objects.search_group_by_keyword(search_query_param)
+        given_study_group_queryset = search_response.context[study_group_context_name]
+
+        assert set(expected_study_group_queryset) == set(given_study_group_queryset)
+
+    @pytest.mark.parametrize("search_query_param, users_context_name", [("", "users")])
```
Comment (Omeramsc) [inline tests/main/test_search_page.py:42]: why do we only look at en empty search query?

File: tests/main/test_search_page.py [hunk 4]
```diff
@@ -0,0 +1,60 @@
...
+    def test_search_context_equals_expected_queryset_users(self, make_search_http_response_from_query_param,
+                                                           search_query_param, users_context_name):
+        search_response = make_search_http_response_from_query_param(search_query_param)
+
+        expected_users_queryset = Profile.search_users_by_keyword(search_query_param)
+        given_users_queryset = search_response.context[users_context_name]
+
+        assert set(expected_users_queryset) == set(given_users_queryset)
+
+    @pytest.mark.parametrize("search_query_param, courses_context_name", [("", "teacher_courses")])
```
Comment (Omeramsc) [inline tests/main/test_search_page.py:52]: why do we only look at en empty search query?

**Review by Omeramsc — CHANGES_REQUESTED** [review]
Nice change, but I can see some missing test cases, try to broad them

File: users/tests.py [hunk 5]
```diff
@@ -83,3 +84,18 @@ def test_filter_by_account_type(self, make_profile, user_idx, account_type, seco
...
         profiles_by_account_type = Profile.filter_by_account_type(new_profile.account_type)
         assert sorted(list(profiles_by_account_type)) == [new_profile]
+
+    @pytest.mark.parametrize("keyword_only_relevant_for_first_profile, account_type, num_of_profiles",
+                             [("first_name0", Account_type.STUDENT, 2),
+                              ("last_name0", Account_type.STUDENT, 2),
+                              ("last_name0", Account_type.STUDENT, 2)])
+    def test_search_users_by_keyword(self, make_profile, keyword_only_relevant_for_first_profile,
+                                     account_type, num_of_profiles):
+        profile_list = [make_profile(i, account_type) for i in range(num_of_profiles)]
```
[the remark this answers is not in the document]
Comment (idan1356) [reply users/tests.py:94]: This test uses the `make_profile` fixture factory, so each profile generated will have unique fields in the format of `fieldname{i}` for `makeprofile(i,..)`
 
What I'm doing here is to create N profiles, then take the first profile with ind=0,
and for each field that is relevant for searching a user by a keyword,
it is asserted that only the first profile is in queryset, for a given `fieldname0` as a keyword. 

Since in this test i use the `make_profile` fixture factory i'm limited to "hardcoded" field names. Would you structure this test differently or add additional test in addition to this one?

File: tests/main/test_search_page.py [hunk 6]
```diff
@@ -0,0 +1,60 @@
...
+        search_response = make_search_http_response_from_query_param(search_query_param)
+        assert search_response.status_code == 200
+        assert 'search.html' in search_response.templates[0].name
+
+    @pytest.mark.parametrize("search_query_param", ["test"])
+    def test_search_view_loaded_unauthorized(self, client, search_query_param):
+        search_response = client.get(f"{reverse('search')}?q={search_query_param}")
+        assert search_response.status_code == 302
+
+    @pytest.mark.parametrize("search_query_param, study_group_context_name", [("", "study_groups")])
```
[the remark this answers is not in the document]
Comment (idan1356) [reply tests/main/test_search_page.py:32]: I will add more test cases

File: tests/main/test_search_page.py (hunk 3 above)
Comment (idan1356) [reply tests/main/test_search_page.py:42]: Will add more test cases

File: tests/main/test_search_page.py (hunk 4 above)
Comment (idan1356) [reply tests/main/test_search_page.py:52]: Will add more test cases

File: users/tests.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (Omeramsc) [reply users/tests.py:94]: I don't understand what you mean by `it is asserted that only the first profile is in queryset`. can you give an example?

and a fixture (or a fixture factory) is a tool to help you, it should never limit you. if it does, you need to create a different fixture (it doesn't cost money :) )

[the remark this answers is not in the document]
Comment (idan1356) [reply users/tests.py:94]: For example for query string `first_name0`, it is asserted that only the first profile (that was created with index 0) is in the queryset (whose `first_name` field is `first_name0`.
As opposed to profiles with indices `i=1,..,N` whose `first_name` field is `first_name1`, `first_name2`,...,`first_nameN`

[the remark this answers is not in the document]
Comment (Omeramsc) [reply users/tests.py:94]: Maybe I'm missing something, so I'll try to clarify again: you check that the search function only finds the exact thing it searched for, meaning that when you search for `first_name0`, you won't find `first_name1` or `first_name` or `first_name01`?
does it cover all those cases? can we prove it?

[the remark this answers is not in the document]
Comment (idan1356) [reply users/tests.py:94]: Yes exactly.

The way I implemented the search function is with `icontains` so case-insensitive containment for the query string test for all the relevant fields in user.
If for example given the 1,...n profiles  made with the make_profile factory i search for `first_name` or `First_NAME` i should get all the profiles in queryset.

Maybe I should also add two more tests to test case insensitive and string containment behavior?

File: tests/main/test_search_page.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (idan1356) [reply tests/main/test_search_page.py]: fixed

File: tests/main/test_search_page.py (hunk 3 above)
Comment (idan1356) [reply tests/main/test_search_page.py]: fixed

File: tests/main/test_search_page.py (hunk 4 above)
Comment (idan1356) [reply tests/main/test_search_page.py]: fixed

Comment (idan1356) [reply tests/main/test_search_page.py]: fixed

File: users/tests.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (Omeramsc) [reply users/tests.py:94]: With tests for search or regex functions, you have to try and fail the test as much as possible, with multiple edge-cases.
So I'll test things like: Case-sensetive, typo, 1 character less than expected, 1 character more than expected. 

If time allows, try to achieve it today. it is important.

[the remark this answers is not in the document]
Comment (idan1356) [reply users/tests.py:93]: Fixed

**Review by Omeramsc — APPROVED** [review]
LGTM

File: users/tests.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (Omeramsc) [reply users/tests.py:93]: Great work! well done
